File: concept/type/EntityTypeImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {EntityType as EntityTypeProto} from "typedb-protocol/proto/concept";
import {Entity} from "../../api/concept/thing/Entity";
import {EntityType} from "../../api/concept/type/EntityType";
import {TypeDBTransaction} from "../../api/connection/TypeDBTransaction";
import {RequestBuilder} from "../../common/rpc/RequestBuilder";
import {Stream} from "../../common/util/Stream";
import {EntityImpl, ThingTypeImpl} from "../../dependencies_internal";
import {Concept} from "../../api/concept/Concept";
import Transitivity = Concept.Transitivity;

export class EntityTypeImpl extends ThingTypeImpl implements EntityType {
    constructor(name: string, root: boolean, abstract: boolean) {
        super(name, root, abstract);
    }

    protected get className(): string {
        return "EntityType";
    }

    isEntityType(): boolean {
        return true;
    }

    asEntityType(): EntityType {
        return this;
    }

    async isDeleted(transaction: TypeDBTransaction): Promise<boolean> {
        return !(await transaction.concepts.getEntityType(this.label.name));
    }

    async create(transaction: TypeDBTransaction): Promise<Entity> {
        const res = await this.execute(transaction, RequestBuilder.Type.EntityType.createReq(this.label));
        return EntityImpl.ofEntityProto(res.entity_type_create_res.entity);
    }

    async getSupertype(transaction: TypeDBTransaction): Promise<EntityType> {
        const res = await this.execute(transaction, RequestBuilder.Type.EntityType.getSupertypeReq(this.label));
        return EntityTypeImpl.ofEntityTypeProto(res.entity_type_get_supertype_res.entity_type);
    }

    async setSupertype(transaction: TypeDBTransaction, superEntityType: EntityType): Promise<void> {
        await this.execute(transaction, RequestBuilder.Type.EntityType.setSupertypeReq(this.label, EntityType.proto(superEntityType)));
    }

    getSupertypes(transaction: TypeDBTransaction): Stream<EntityType> {
        return this.stream(transaction, RequestBuilder.Type.EntityType.getSupertypesReq(this.label)).flatMap(
            resPart => Stream.array(resPart.entity_type_get_supertypes_res_part.entity_types)
        ).map(EntityTypeImpl.ofEntityTypeProto);
    }

    getSubtypes(transaction: TypeDBTransaction): Stream<EntityType>;
    getSubtypes(transaction: TypeDBTransaction, transitivity: Transitivity): Stream<EntityType>;
    getSubtypes(transaction: TypeDBTransaction, transitivity?: Transitivity): Stream<EntityType> {
        if (!transitivity) transitivity = Transitivity.TRANSITIVE;
        return this.stream(transaction, RequestBuilder.Type.EntityType.getSubtypesReq(this.label, transitivity.proto())).flatMap(
            resPart => Stream.array(resPart.entity_type_get_subtypes_res_part.entity_types)
        ).map(EntityTypeImpl.ofEntityTypeProto);
    }

    getInstances(transaction: TypeDBTransaction): Stream<Entity>;
    getInstances(transaction: TypeDBTransaction, transitivity: Transitivity): Stream<Entity>;
    getInstances(transaction: TypeDBTransaction, transitivity?: Transitivity): Stream<Entity> {
        if (!transitivity) transitivity = Transitivity.TRANSITIVE;
        return this.stream(transaction, RequestBuilder.Type.EntityType.getInstancesReq(this.label, transitivity.proto())).flatMap(
            resPart => Stream.array(resPart.entity_type_get_instances_res_part.entities)
        ).map(EntityImpl.ofEntityProto);
    }
}

export namespace EntityTypeImpl {
    export function ofEntityTypeProto(proto: EntityTypeProto): EntityType {
        if (!proto) return null;
        return new EntityTypeImpl(proto.label, proto.is_root, proto.is_abstract);
    }
}

File: concept/type/EntityTypeImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.type;

import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.type.EntityType;
import com.vaticle.typedb.client.concept.ConceptManagerImpl;
import com.vaticle.typedb.client.concept.thing.EntityImpl;
import com.vaticle.typedb.client.jni.Transitivity;

import javax.annotation.Nullable;
import java.util.stream.Stream;

import static com.vaticle.typedb.client.jni.typedb_client.entity_type_create;
import static com.vaticle.typedb.client.jni.typedb_client.entity_type_get_instances;
import static com.vaticle.typedb.client.jni.typedb_client.entity_type_get_subtypes;
import static com.vaticle.typedb.client.jni.typedb_client.entity_type_get_supertype;
import static com.vaticle.typedb.client.jni.typedb_client.entity_type_get_supertypes;
import static com.vaticle.typedb.client.jni.typedb_client.entity_type_set_supertype;

public class EntityTypeImpl extends ThingTypeImpl implements EntityType {
    public EntityTypeImpl(com.vaticle.typedb.client.jni.Concept concept) {
        super(concept);
    }

    @Override
    public final EntityImpl create(TypeDBTransaction transaction) {
        return new EntityImpl(entity_type_create(((ConceptManagerImpl) transaction.concepts()).transaction, concept));
    }

    @Override
    public final void setSupertype(TypeDBTransaction transaction, EntityType entityType) {
        entity_type_set_supertype(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((EntityTypeImpl) entityType).concept);
    }

    @Nullable
    @Override
    public EntityTypeImpl getSupertype(TypeDBTransaction transaction) {
        com.vaticle.typedb.client.jni.Concept res = entity_type_get_supertype(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
        if (res != null) return new EntityTypeImpl(res);
        else return null;
    }

    @Override
    public final Stream<EntityTypeImpl> getSupertypes(TypeDBTransaction transaction) {
        return entity_type_get_supertypes(((ConceptManagerImpl) transaction.concepts()).transaction, concept).stream().map(EntityTypeImpl::new);
    }

    @Override
    public final Stream<EntityTypeImpl> getSubtypes(TypeDBTransaction transaction) {
        return entity_type_get_subtypes(((ConceptManagerImpl) transaction.concepts()).transaction, concept, Transitivity.Transitive).stream().map(EntityTypeImpl::new);
    }

    @Override
    public final Stream<EntityTypeImpl> getSubtypesExplicit(TypeDBTransaction transaction) {
        return entity_type_get_subtypes(((ConceptManagerImpl) transaction.concepts()).transaction, concept, Transitivity.Explicit).stream().map(EntityTypeImpl::new);
    }

    @Override
    public final Stream<EntityImpl> getInstances(TypeDBTransaction transaction) {
        return entity_type_get_instances(((ConceptManagerImpl) transaction.concepts()).transaction, concept, Transitivity.Transitive).stream().map(EntityImpl::new);
    }

    @Override
    public final Stream<EntityImpl> getInstancesExplicit(TypeDBTransaction transaction) {
        return entity_type_get_instances(((ConceptManagerImpl) transaction.concepts()).transaction, concept, Transitivity.Explicit).stream().map(EntityImpl::new);
    }
}
